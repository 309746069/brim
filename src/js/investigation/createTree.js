/* @flow */
import {isEqual} from "lodash"

import type {Investigation} from "../reducers/investigation"
import Tree from "../models/Tree"

export function createInvestigationTree(investigation: Investigation) {
  let tree = new Tree({data: "ROOT", children: []})

  for (let finding of investigation) {
    let node = tree.getRoot()

    eachFilter(finding.record, (filter) => {
      if (!node) return
      let nextNode = node.children.find((child) =>
        isEqual(child.data.filter, filter)
      )
      if (nextNode) {
        node = nextNode
      } else {
        node = node.addChild({filter, finding})
      }
    })
  }

  return tree
}

function eachFilter(searchBar, callback) {
  const {pins, program} = searchBar
  const filters = [...pins]

  if (!/^\s*$/.test(program)) filters.push(program)

  return filters.forEach(callback)
}
