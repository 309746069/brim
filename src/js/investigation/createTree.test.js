/* @flow */
import {createInvestigationTree} from "./createTree"

function search(pins, program) {
  return {
    ts: new Date(0),
    search: {
      pins,
      program,
      space: "default",
      span: [new Date(1), new Date(2)]
    }
  }
}

test("Remove all pins and make a new current", () => {
  let tree = createInvestigationTree([
    search(["a", "b"], "c"),
    search(["a", "b"], "d"),
    search(["a", "b"], "d"),
    search(["a"], "d"),
    search([], "e")
  ])

  expect(tree.toJSON()).toMatchSnapshot()
})
