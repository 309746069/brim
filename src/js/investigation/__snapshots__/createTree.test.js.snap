// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove all pins and make a new current 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "data": Object {
                "filter": "c",
                "finding": Object {
                  "search": Object {
                    "pins": Array [
                      "a",
                      "b",
                    ],
                    "program": "c",
                    "space": "default",
                    "span": Array [
                      1970-01-01T00:00:00.001Z,
                      1970-01-01T00:00:00.002Z,
                    ],
                  },
                  "ts": 1970-01-01T00:00:00.000Z,
                },
              },
            },
            Object {
              "children": Array [],
              "data": Object {
                "filter": "d",
                "finding": Object {
                  "search": Object {
                    "pins": Array [
                      "a",
                      "b",
                    ],
                    "program": "d",
                    "space": "default",
                    "span": Array [
                      1970-01-01T00:00:00.001Z,
                      1970-01-01T00:00:00.002Z,
                    ],
                  },
                  "ts": 1970-01-01T00:00:00.000Z,
                },
              },
            },
          ],
          "data": Object {
            "filter": "b",
            "finding": Object {
              "search": Object {
                "pins": Array [
                  "a",
                  "b",
                ],
                "program": "c",
                "space": "default",
                "span": Array [
                  1970-01-01T00:00:00.001Z,
                  1970-01-01T00:00:00.002Z,
                ],
              },
              "ts": 1970-01-01T00:00:00.000Z,
            },
          },
        },
        Object {
          "children": Array [],
          "data": Object {
            "filter": "d",
            "finding": Object {
              "search": Object {
                "pins": Array [
                  "a",
                ],
                "program": "d",
                "space": "default",
                "span": Array [
                  1970-01-01T00:00:00.001Z,
                  1970-01-01T00:00:00.002Z,
                ],
              },
              "ts": 1970-01-01T00:00:00.000Z,
            },
          },
        },
      ],
      "data": Object {
        "filter": "a",
        "finding": Object {
          "search": Object {
            "pins": Array [
              "a",
              "b",
            ],
            "program": "c",
            "space": "default",
            "span": Array [
              1970-01-01T00:00:00.001Z,
              1970-01-01T00:00:00.002Z,
            ],
          },
          "ts": 1970-01-01T00:00:00.000Z,
        },
      },
    },
    Object {
      "children": Array [],
      "data": Object {
        "filter": "e",
        "finding": Object {
          "search": Object {
            "pins": Array [],
            "program": "e",
            "space": "default",
            "span": Array [
              1970-01-01T00:00:00.001Z,
              1970-01-01T00:00:00.002Z,
            ],
          },
          "ts": 1970-01-01T00:00:00.000Z,
        },
      },
    },
  ],
  "data": "ROOT",
}
`;
